sacc/data_types.py

```diff
@@ -0,0 +1,374 @@
+import warnings
+
+from astropy.table import Table
+
+from .utils import Namespace, hide_null_values, null_values, camel_case_split_and_lowercase
+
+# The format for a data type name looks like this:
+# {sources}_{properties}_{statistic_type}[_{statistic_subtype}]
+#     sources: type(s) of astrophysical sources to which this applies
+#     properties: feature(s)/characterisic(s) of those sources/fields to which the statistic applies
+#     statistic_type: mathematical type of the statistic
+#     statistic_subtype: optional additional specifier
+
+required_tags = {
+
+    "galaxy_density_xi": ["theta"],
+    "galaxy_density_cl": ["ell"],
+    "galaxy_density_cosebi": [],
+
+    "galaxy_shear_cl_ee": ["ell"],
+    "galaxy_shear_cl_bb": ["ell"],
+    "galaxy_shear_cl_ee": ["ell"],
+    "galaxy_shear_cl_bb": ["ell"],
+    "galaxy_shear_cl_eb": ["ell"],
+    "galaxy_shear_cl_be": ["ell"],
+    "galaxy_shear_xi_plus": ["theta"],
+    "galaxy_shear_xi_minus": ["theta"],
+    "galaxy_shear_xi_eplus": ["theta"],
+    "galaxy_shear_xi_eminus": ["theta"],
+    "galaxy_shear_xi_bplus": ["theta"],
+    "galaxy_shear_xi_bminus": ["theta"],
+    "galaxy_shear_cosebi_ee": [],
+    "galaxy_shear_cosebi_bb": [],
+
+    "galaxy_convergence_cl": ["ell"],
+    "galaxy_convergence_xi": ["theta"],
+
+    "galaxy_shearDensity_cl_e": ["ell"],
+    "galaxy_shearDensity_cl_b": ["ell"],
+    "galaxy_shearDensity_xi_t": ["theta"],
+    "galaxy_shearDensity_xi_x": ["theta"],
+    "galaxy_shearDensity_cosebi_e": [],
+    "galaxy_shearDensity_cosebi_b": [],
+
+
+    "cmb_temperature_cl_tt": ["ell"],
+
+    "cmb_polarization_cl_ee": ["ell"],
+    "cmb_polarization_cl_bb": ["ell"],
+    "cmb_polarization_cl_eb": ["ell"],
+    "cmb_polarization_cl_be": ["ell"],
+
+    "cmb_temperaturePolarization_cl_t": ["ell"],
+    "cmb_temperaturePolarization_cl_t": ["ell"],
+    "cmbGalaxy_temperatureShear_cl_te": ["ell"],
+    "cmbGalaxy_temperatureShear_cl_tb": ["ell"],
+
+    "cluster_mass_count_wl": [],
+    "cluster_mass_count_xray": [],
+
+    "clusterGalaxy_densityShear_cl_te": ["ell"],
+    "clusterGalaxy_densityShear_cl_tb": ["ell"],
+    "clusterGalaxy_densityShear_xi_t": ["theta"],
+    "clusterGalaxy_densityShear_xi_x": ["theta"],
+    "clusterGalaxy_densityShear_xi_rt": ["radius"],
+    "clusterGalaxy_densityShear_xi_rx": ["radius"],
```

Comment:
So here's a list of missing tags (`<...>` means all possible cl/xi/cosebi):
```
galaxy_convergenceDensity_<...>
clusterGalaxy_density_<...>
cmbGalaxy_temperatureDensity_<...>
cmbGalaxy_polarizationDensity_<...>
galaxy_shearConvergence_<...>
clusterGalaxy_densityShear_cosebi<...>
cmbGalaxy_temperatureShear_<xi/cosebi...>
cmbGalaxy_polarizationShear_<...>
clusterGalaxy_densityConvergence_<...>
cmbGalaxy_temperatureConvergence_<...>
cmbGalaxy_polarizationConvergence_<...>
cluster_density_<...>
cmbCluster_temperatureDensity_<...>
cmbCluster_polarizationDensity_<...>
cmb_temperaturePolarization_<xi/cosebi>
cmb_polarization_<xi/cosebi>
```
Some of these have zero expectation value, but can be null tests (the same way that you have shear_eb, for example). Others are not mainstream (e.g. CMB correlation functions), but they can definitely be computed. Others are more usual than some of the ones you've added (e.g. I'd argue that `cmbGalaxy_temperatureDensity` is more likely to be interesting than `cmbGalaxy_temperatureShear`).

A few questions:
- do `galaxy_density_cosebi`s make sense (same for convergence, cluster_density, cmb_temperature etc.)?
- what do `xi_eplus` and `xi_bplus` mean? and what's the difference with `xi_plus` (same for `minus`)?
- should there be a `cmb_convergence` type for CMB lensing? That cross-correlation seems to me more important than the CMB temperature/polarization ones.
- why do you have  `galaxy_shearDensity_cl_e` but `clusterGalaxy_densityShear_cl_te`?
- what is `clusterGalaxy_densityShear_xi_rt`? what is the difference with `clusterGalaxy_densityShear_xi_t`?
- `cmb_temperaturePolarization_cl_t` is repeated twice
- You can have `cmbGalaxy` but not `galaxyCmb`  (and many similar others). Will this make it confusing?

I think there should also be some guidance for what the user should do if they want to include data points that are not supported here. E.g. for systematics checks we will definitely want to compute galaxy-star correlation functions, or galaxy-systematic correlation functions (where "systematic" is any diagnostic map, same for shear).

Reply:
Many thanks!  I've added the missing ones here.

> do galaxy_density_cosebis make sense (same for convergence, cluster_density, cmb_temperature etc.)?

- No!  Deleted

> what do xi_eplus and xi_bplus mean? and what's the difference with xi_plus (same for minus)?

- I was thinking of the E-only and B-only xi values.  But they're rarely useful - I've deleted.

> should there be a cmb_convergence type for CMB lensing? That cross-correlation seems to me more important than the CMB temperature/polarization ones.

- added, thanks!

> why do you have galaxy_shearDensity_cl_e but clusterGalaxy_densityShear_cl_te?

- changed to canonical order 

> what is clusterGalaxy_densityShear_xi_rt? what is the difference with clusterGalaxy_densityShear_xi_t?
- I was thinking of xi(r) vs xi(theta).  But not useful or clear I think - deleted and have put an example of doing this in a notebook

> cmb_temperaturePolarization_cl_t is repeated twice

- thanks, deleted

> You can have cmbGalaxy but not galaxyCmb (and many similar others). Will this make it confusing?

- I think better to pick one of the two?

> I think there should also be some guidance for what the user should do if they want to include data points that are not supported here

- good point - added to the docs.